toolchain/check/impl.cpp

```diff
@@ -285,15 +231,54 @@ auto AddConstantsToImplWitnessFromConstraint(Context& context,
 
   // For each non-function associated constant, update witness entry.
   for (auto index : llvm::seq(assoc_entities.size())) {
-    auto decl_id = assoc_entities[index];
-    decl_id =
-        context.constant_values().GetInstId(SemIR::GetConstantValueInSpecific(
-            context.sem_ir(), interface_type->specific_id, decl_id));
+    auto decl_id =
+        context.constant_values().GetConstantInstId(assoc_entities[index]);
     CARBON_CHECK(decl_id.has_value(), "Non-constant associated entity");
     if (auto decl =
             context.insts().TryGetAs<SemIR::AssociatedConstantDecl>(decl_id)) {
+      const auto& assoc_const =
+          context.associated_constants().Get(decl->assoc_const_id);
       auto& witness_value = witness_block[index];
       auto rewrite_value = rewrite_values[index];
+
+      // If the associated constant has a symbolic type, convert the rewrite
+      // value to that type now we know the value of `Self`.
+      SemIR::TypeId assoc_const_type_id = decl->type_id;
+      if (context.types().GetConstantId(assoc_const_type_id).is_symbolic()) {
+        // Form a specific for the associated constant's generic.
+        SemIR::SpecificId assoc_const_specific_id =
+            GetSelfSpecificForInterfaceMemberWithSelfType(
+                context, impl.constraint_id, interface_type->specific_id,
+                assoc_const.generic_id,
+                context.GetTypeIdForTypeInst(impl.self_id), witness_id);
+        assoc_const_type_id = SemIR::GetTypeInSpecific(
+            context.sem_ir(), assoc_const_specific_id, assoc_const_type_id);
```

Comment:
Reminder to use the combined GetAssociatedEntityType function.

Reply:
Done.